### ISSUE TITLE
[CSS UI-4] caret-animation property

### ISSUE THREAD

**Issue by schenney-chromium** [mozilla/standards-positions#1100]
## Request for Mozilla Position on an Emerging Web Specification

* Specification title: CSS UI 4
* Specification or proposal URL (if available): https://drafts.csswg.org/css-ui-4/#caret-animation
* Explainer URL (if available): See the [CSS WG Issue](https://github.com/w3c/csswg-drafts/issues/9707)
* Proposal author(s) (`@`-mention GitHub accounts): [frivoal](https://github.com/frivoal)
* Caniuse.com URL (optional): 
* Bugzilla URL (optional): 
* Mozillians who can provide input (optional): 
* WebKit standards-position: Pending

### Other information
This is already in the spec after the resolution of [CSS #9707]( https://github.com/w3c/csswg-drafts/issues/9707)
My recollection from the CSS WG meeting is that Emilio was OK but the log does not reflect that so I may well be confusing issues.
Chromium intent: https://chromestatus.com/feature/5082469066604544

**Comment by emilio** [comment]
In general this seems sensible to me, though I'm curious if a11y folks have concerns about this? cc @jcsteh

It should be relatively easy to implement (i.e., Firefox supports `ui.caretBlinkTime=-1` to prevent the caret animation).

A few questions I have (and the spec doesn't quite determine unless I'm missing something):

 * What happens when you have the caret between element boundaries? i.e. if you have `<span style="caret-animation: manual">foo</span><span style="caret-animation: auto">bar</span>`, is it well defined which one applies if you're in between the last `o` and the `b`? Is it the intention that you take the caret-animation value from the closest editing host (and thus the value on individual `<span>`s is meaningless?). That's my read from the "applies to" line... But the spec could be more explicit.

 * If you apply this property dynamically, When are UAs expected to honor it? Immediately? When the caret moves? Sometime else?

 * The spec doesn't seem to consider non-editable caret (like caret browsing which IIRC Chromium also has). Should it?

If you agree these are not defined, it'd be great to open issues about them...

**Comment by jcsteh** [comment]
> The spec text mentions this being a a11y benefit for users impacted by motion.

It could equally be an a11y detriment, though, if the user has very specific OS or browser settings for blink time, etc. and some site goes and overrides those. Given that certain kinds of blinking can cause severe physical reactions for some folks, this might not be a minor concern. That said, the fact that you can already animate caret-color means this is already possible, though we shouldn't necessarily dig the hole deeper just because the hole has already been dug. :) I guess browsers could choose to ignore this if commanded by the user, just as they can behave specially with forced colors, for example. Does that seem reasonable?

CC @MReschenberg, who has deeper experience and understanding of colour and motion issues than I, in case she has further thoughts.

**Comment by emilio** [comment]
Yeah so... Caret-color should be handled like other colors in forced-colors mode, probably. And that means that caret-animation: none would unintentionally disable the custom blinking animation altogether for those users which seems unfortunate. It's at least a bit footgunny

**Comment by schenney-chromium** [comment]
Thanks for the input.

Note the values are `caret-animation: auto` (browser does what it will) and `caret-animation: manual` intended to mean the site is manually animating the caret. Which leads me to propose this alternative.

The spec could be changed to say that if you set `caret-animation: manual` it will only apply if you are actually manually animating the caret. i.e. if you have `caret-color` set to an animated color. I don't have enough spec experience to know if such a thing would be acceptable, but it could be implemented.

And the a11y use case could be delegated to the browser and/or under the control of `prefers-reduced-motion`.

I agree that caret-color should be modified by forced-colors.

Thoughts?

A couple of other things:
* There is no MDN page for caret-animation. I'll work on fixing that once we nail down mutually agreeable behavior.
* `caret-color: transparent` is supported by all browsers which is definitely not good for users if used improperly but is currently used when sites totally override the caret styling by implementing their own caret.

**Comment by schenney-chromium** [comment]
Update: There's a spec PR to clarify when the CSS caret properties apply: https://github.com/w3c/csswg-drafts/pull/11373

Any update on the Mozilla side?

**Comment by schenney-chromium** [comment]
As I said above, there is a [spec PR](https://github.com/w3c/csswg-drafts/pull/11373) up to say that caret-* properties only apply to text input elements. Emilio is active on reviewing that and it should address both the source of the properties for styling the caret and when it applies at all. I'll also add a comment on that PR about when dynamic updates should apply. 

The spec already allows UAs to override the property if they want, so I agree that user browser settings should take precedence over page authors.

I hope to ship this in Chromium in the coming weeks, once the spec changes land.
